services/management/backup/backups_service.go

```diff
@@ -18,116 +18,356 @@ package backup
+			serviceCh <- s
+		}
+
 		return nil
 	})
-	if errTX != nil {
-		return nil, errTX
+
+	// check each service for compatibility
+	compatibleServicesMap := make(map[string]struct{}, len(filteredServices))
+	m := &sync.Mutex{}
+	for i := 0; i < workers; i++ {
+		g.Go(func() error {
+			for service := range serviceCh {
+				pmmAgents, err := models.FindPMMAgentsForService(s.db.Querier, service.ServiceID)
+				if err != nil {
+					return err
+				}
+
+				if len(pmmAgents) == 0 {
+					continue
+				}
+
+				agentID := pmmAgents[0].AgentID
+
+				dbVersion, err := s.versionService.GetLocalMySQLVersion(agentID)
+				if err != nil {
+					s.l.WithError(err).Infof("skip incompatible service with id %q, agent id %q: "+
+						"cannot get local MySQL version", service.ServiceID, agentID)
+					continue
+				}
+
+				if artifactDBVersion != dbVersion {
+					s.l.WithError(err).Infof("skip incompatible service with id %q, agent id %q: "+
+						"version mismatch, artifact version %q != db version %q", service.ServiceID, agentID, artifactDBVersion, dbVersion)
+					continue
+				}
+
+				toolVersion, err := s.versionService.GetXtrabackupVersion(agentID)
+				if err != nil {
+					s.l.WithError(err).Infof("skip incompatible service with id %q, agent id %q: "+
+						"cannot get xtrabackup version", service.ServiceID, agentID)
+					continue
+				}
+
+				_ = toolVersion
+				// TODO: check compatibility between db and tool versions
+
+				m.Lock()
+				compatibleServicesMap[service.ServiceID] = struct{}{}
+				m.Unlock()
+			}
+
+			return nil
+		})
 	}
 
-	locationConfig := &models.BackupLocationConfig{
-		PMMServerConfig: location.PMMServerConfig,
-		PMMClientConfig: location.PMMClientConfig,
-		S3Config:        location.S3Config,
+	if err := g.Wait(); err != nil {
+		return nil, err
 	}
 
-	switch svc.ServiceType {
+	compatibleServices := make([]*models.Service, 0, len(compatibleServicesMap))
+	for _, service := range filteredServices {
+		if _, ok := compatibleServicesMap[service.ServiceID]; ok {
+			compatibleServices = append(compatibleServices, service)
+		}
+	}
+
+	return compatibleServices, nil
+}
+
+// ListServicesForRestore lists compatible service for restoring given artifact.
+func (s *BackupsService) ListServicesForRestore(
+	_ context.Context,
+	req *backupv1beta1.ListServicesForRestoreRequest,
+) (*backupv1beta1.ListServicesForRestoreResponse, error) {
+	artifact, err := models.FindArtifactByID(s.db.Querier, req.ArtifactId)
+	if err != nil {
+		return nil, err
+	}
+
+	serviceType := models.ServiceType(artifact.Vendor)
+	switch serviceType {
 	case models.MySQLServiceType:
-		err = s.jobsService.StartMySQLBackupJob(job.ID, job.PMMAgentID, 0, req.Name, config, locationConfig)
-	case models.MongoDBServiceType:
-		err = s.jobsService.StartMongoDBBackupJob(job.ID, job.PMMAgentID, 0, req.Name, config, locationConfig)
-	case models.PostgreSQLServiceType,
+	case models.MongoDBServiceType, // TODO: add mongo support
```

Comment:
let's return only service where backup made from for now for MongoDB.

Reply:
returning single service for mongo